packages/grafana-data/src/field/fieldOverrides.ts

```diff
@@ -97,33 +97,38 @@ export function applyFieldOverrides(options: ApplyFieldOverrideOptions): DataFra
     }
   }
 
-  return options.data.map((frame, index) => {
+  return options.data.map((originalFrame, index) => {
     // Need to define this new frame here as it's passed to the getLinkSupplier function inside the fields loop
-    const newFrame: DataFrame = { ...frame };
+    const newFrame: DataFrame = { ...originalFrame };
+    // Copy fields
+    newFrame.fields = newFrame.fields.map((field) => {
+      return {
+        ...field,
+        config: cloneDeep(field.config),
```

Comment:
Assuming this runs on everything.... perhaps we should clone the whole DataFrame before PanelQueryRunner starts applying changes, and then everything in the pipeline can safely mutate.  With the current model, we have a few deep clones scattered about adding a bit of complexity.

cc @leeoniya

Reply:
yea this deepClone was new to me, was added recently. But we do not want to deep clone the values! so I think being more deliberate about what is shallow and what is deep like above I think is ok

Reply:
if it's an array of plain values then a "deep" clone on it is a cheap `values.slice()` call that i assume has copy-on-write semantics in V8. the trick is not to iterate those arrays and do a `push` loop, which may need custom handling to avoid in cloneDeep(). i think cloning the rest of the structures will be much more expensive.

Reply:
Been looking at this, I would like to cleanup the code in PanelQueryRunner.getData (that handles copy field config from last result, etc), there is a lot of copying the data frame there and then again in applyFieldOverrides. The only problem is applyFieldOverrides is also used elsewhere. So making applyFieldOverrides fully mutable could have other effects there.